tests/stream_states.rs

```diff
@@ -808,3 +808,65 @@ fn send_data_after_headers_eos() {
 fn exceed_max_streams() {
 }
 */
+
+
+#[test]
+fn rst_while_closing() {
+    // Test to reproduce panic in issue #246 --- receipt of a RST_STREAM frame
+    // on a stream in the Half Closed (remote) state with a queued EOS causes
+    // a panic.
+    let _ = ::env_logger::try_init();
+    let (io, srv) = mock::new();
+
+    let srv = srv.assert_client_handshake()
+        .unwrap()
+        .recv_settings()
+        .recv_frame(
+            frames::headers(1)
+                .request("GET", "https://example.com/")
+        )
+        .send_frame(frames::headers(1).response(200))
+        .send_frame(frames::headers(1).eos())
+        // Idling for a moment here is necessary to ensure that the client
+        // enqueues its TRAILERS frame *before* we send the RST_STREAM frame
+        // which causes the panic.
+        .idle_ms(1)
```

Comment:
Is this racey at all? Maybe we need a way to tell the mocks to wait on a future, to allow for synchronization?

Reply:
It could be potentially racy; changing the mocks to allow waiting on a future seems like a better approach long-term.